### PR TITLE
utils: resolve_binary_path: Support symbolic links in ENV

### PR DESCRIPTION
**Pull request by Rtoax**
On some systems `/bin` is a symbolic link to `/usr/bin` (`/bin -> /usr/bin`), When using the '`-c`' parameter to specify the executable program in the environment variable, the problem of "multiple binary files matching" occurs, for example:

```
  $ sudo bpftrace -e 'kprobe:do_nanosleep { printf("%d sleeping\n", pid); }' -c 'sleep 5'
  ERROR: Failed to fork child: path 'sleep' must refer to a unique binary but matched 2 binaries
```

Debugging the bpftrace program through gdb can also be verified:

```
  (gdb) p valid_executable_paths
  $9 = std::vector of length 2, capacity 2 = {"/bin/ls", "/usr/bin/ls"}
```

This commit fixes the issue.

### CODE REVIEW

File: src/utils.cpp [hunk 1]
```diff
@@ -849,10 +857,17 @@ resolve_binary_path(const std::string &cmd, const char *env_paths, int pid)
...
+      rel_path = realpath(path);
+
     if (bcc_elf_is_exe(rel_path.c_str()) ||
         bcc_elf_is_shared_obj(rel_path.c_str()))
-      valid_executable_paths.push_back(rel_path);
+    {
+      auto it = std::find(valid_executable_paths.begin(),
+                          valid_executable_paths.end(), rel_path);
+      if (it == valid_executable_paths.end()) {
+        valid_executable_paths.push_back(rel_path);
```
Comment (viktormalik) [inline src/utils.cpp:868]: Instead of doing this manually, make `valid_executable_paths` a `std::unordered_set`.

File: src/utils.cpp [hunk 2]
```diff
@@ -308,6 +308,14 @@ std::vector<std::string> split_string(const std::string &str,
   return elems;
 }
 
+std::string realpath(std::string const& path) {
+  int exist = ::access(path.c_str(), F_OK);
+  if (exist != 0) {
```
Comment (viktormalik) [inline src/utils.cpp:313]: nit: [access](https://man7.org/linux/man-pages/man2/access.2.html) returns -1 on error so let's check it directly:
```suggestion
  if (exist == -1) {
```

**Review by viktormalik — COMMENTED** [review]
Looks like a useful feature, thanks!

Please, look at the CI failures and fix clang-format issues.

File: src/utils.cpp (hunk 2 above)
Comment (Rtoax) [reply src/utils.cpp]: Fix it , thanks.

File: src/utils.cpp [hunk 3]
```diff
@@ -842,19 +852,24 @@ resolve_binary_path(const std::string &cmd, const char *env_paths, int pid)
   if (cmd.find("*") != std::string::npos)
     candidate_paths = ::expand_wildcard_paths(candidate_paths);
 
+  std::unordered_set<std::string> abs_path;
```
Comment (viktormalik) [inline src/utils.cpp:855]: Why can't we just change `valid_executable_paths` (and consequently the return type of this function) to `std::unordered_set` and avoid the copy below?

**Review by viktormalik — COMMENTED** [review]
This seems to break the tests in several ways. Probe listing tests are broken b/c abs path is reported now. Also several tests are now segfaulting.

You may need to take another approach to resolve this issue.

File: src/utils.cpp (hunk 3 above)
Comment (Rtoax) [reply src/utils.cpp:855]: It's wired, On my `CentOS-Stream9`, there are no failed tests at all.

```
[----------] Global test environment tear-down
[==========] 401 tests from 22 test suites ran. (610238 ms total)
[  PASSED  ] 401 tests.
```

I'll try another approach, maybe submit new V2 patch, thanks.

**Comment by Rtoax** [comment]
@viktormalik I just close this PR, and start V2 (https://github.com/iovisor/bpftrace/pull/2689)

**Pull request closed without merging**